server/enclave/ecalls.cpp

```diff
@@ -0,0 +1,135 @@
+{
+    size_t encryption_metadata_length = 3;
+
+    unsigned char* encrypted_old_params_cpy[encryption_metadata_length];
+    size_t lengths[] = {old_params_length, CIPHER_IV_SIZE, CIPHER_TAG_SIZE};
+    copy_arr_to_enclave(encrypted_old_params_cpy, 
+            encryption_metadata_length, 
+            encrypted_old_params, 
+            lengths);
+
+    unsigned char serialized_old_params[old_params_length];
+    decrypt_bytes(*encrypted_old_params_cpy, 
+            *(encrypted_old_params + 1), 
+            *(encrypted_old_params + 2), 
+            old_params_length,
+            (unsigned char**) &serialized_old_params);
+
+    map<string, vector<double>> params = deserialize(string((const char*) serialized_old_params));
+
+    
+    vector<map<string, vector<double>>> accumulator;
+    set<string> vars_to_aggregate;
+
+    for (int i = 0; i < accumulator_length; i++) {
+        unsigned char decrypted_accumulator[accumulator_lengths[i]];
+        decrypt_bytes(*encrypted_accumulator[i],
+                *(encrypted_accumulator[i] + 1),
+                *(encrypted_accumulator[i] + 2),
+                accumulator_lengths[i],
+                (unsigned char**) &decrypted_accumulator);
+
+        map<string, vector<double>> params = deserialize(string((const char*) decrypted_accumulator));
+
+        for (const auto& pair : params) {
+            vars_to_aggregate.insert(pair.first);
+        }
+
+        accumulator.push_back(params);
+    }
+
+    for (string v_name : vars_to_aggregate) {
+        vector<double> n_local_iters;
+        vector<vector<double>> vars;
+
+        for (map<string, vector<double>> acc_params : accumulator) {
+            if (acc_params.find(v_name) == acc_params.end()) { // these params don't have the variable from client
+                continue;
+            }
+
+            // Each params map will have an additional key "_contribution" to hold the number of local iterations
+            double n_iter = acc_params["_contribution"][0];
+            n_local_iters.push_back(n_iter);
+
+            // Weighted using local iterations
+            vector<double>& weights = acc_params[v_name];
+            for_each(weights.begin(), weights.end(), [&n_iter](double& d) { d *= n_iter; });
+            vars.push_back(weights);
+        }
+
+        if (n_local_iters.empty()) {
+            continue; // Didn't receive this variable from any clients
+        }
+
+        vector<double> new_val(vars.size(), 0.0);
+        double iters_sum = accumulate(n_local_iters.begin(), n_local_iters.end(), 0);
+        for (int i = 0; i < new_val.size(); i++) {
+            for (vector<double> weights : vars) {
+                new_val[i] += weights[i];
+            }
+            new_val[i] /= iters_sum;
+        }
+        params[v_name] = new_val;
+    }
+
+    string serialized_new_params = serialize(params);
+
+    unsigned char** encrypted_new_params = new unsigned char*[encryption_metadata_length * sizeof(char*)];
```

Comment:
It might be better to create an unsigned char array (i.e. `unsigned char[]`) of length `serialized_new_params.length() + CIPHER_IV_SIZE + CIPHER_TAG_SIZE` here instead of a `unsigned char**` and pass that into `encrypt_bytes()`. This will of course require changes to `encrypt_bytes()`'s function signature, but I think this will be cleaner. Since you know `serialized_new_params.length()`, `CIPHER_IV_SIZE`, and `CIPHER_TAG_SIZE`, you should be able to properly index into and fetch the ciphertext, the IV, and the tag from this unsigned char array. 

Would also be good to do the same for `decrypt_bytes()`.

Reply:
Actually maybe we can do this in a later PR, assuming that this current code works properly. @podcastinator what do you think?

Reply:
Yeah, I agree that the memory allocation should be handled by the caller. We can improve this later perhaps. I will create an issue for this.